### DISCUSSION TITLE
Creating PWAs

### DISCUSSION THREAD

**Discussion by skeledrew** [elimintz/justpy#289] [Q&A] [answered]
I am an experienced Python developer, and I'd like to create a website with PWA support. How feasible is this with JustPy (particularly re creating service workers)? I am wondering if I may need to use something like [Brython workers](https://brython.info/static_doc/en/worker.html) to bridge the gap.

**Comment by elimintz** [comment, marked as answer]
A main feature of PWAs is that they can work with no or a slow internet connection. JustPy applications cannot since in JustPy there is no frontend/backend distinction. You take care of events in the frontend in the backend basically. A python in the browser solution like Brython may be a better way to go.

If your main concern is long running background tasks and can assume good internet connection, you can use JustPy effectively by running those tasks as part of your application. Please take a look at this example: https://justpy.io/tutorial/pushing_data/#clock
`clock_counter` is a background task in this example

You can initiate your background tasks in the init function and then have them run in the same async loop as the application. Or you can use the say, the multiprocessing module to run them and interact with them using a function you initiate in the init function.

If this is not clear, please let me know.

**Reply by skeledrew** [reply]
I see. My main concern is for users to have a reasonable app-like experience; I think relatively stable connections can be assumed. The reason I emphasized service workers is because for Chrome at least it seems to be mandatory. I'm new to front-end development and only recently started looking into PWAs, so there's a lot I am unsure of. Would doing async or multiprocessing be enough to satisfy that worker requirement?

**Reply by elimintz** [reply]
If there is a stable internet connection, your JustPy app will work well and users will not see any difference. It will not be a PWA, but for the users it won't make a difference even though there are no service workers initiated by the app running in the front end.

**Reply by skeledrew** [reply]
OK, thank you.
